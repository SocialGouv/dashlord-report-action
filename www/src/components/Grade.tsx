import * as React from "react";

import { Badge } from "react-bootstrap";

type GradeProps = { grade: string; label?: string; small?: boolean };

export const Grade: React.FC<GradeProps> = ({ grade, label, small }) => {
  const newGrade = grade.substring(0, 1).toUpperCase();
  const variant =
    newGrade === "F"
      ? "danger"
      : newGrade === "E"
      ? "danger"
      : newGrade === "D"
      ? "warning"
      : newGrade === "C"
      ? "info"
      : newGrade === "B"
      ? "info"
      : newGrade === "A"
      ? "success"
      : "danger";
  return (
    <Badge variant={variant} style={{ fontSize: small ? "1.3em" : "2em" }}>
      {label !== undefined ? label : grade}
    </Badge>
  );
};
