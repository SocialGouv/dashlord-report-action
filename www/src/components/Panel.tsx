import * as React from "react";

import { Card } from "react-bootstrap";
import { ExternalLink } from "react-feather";

type PanelProps = {
  title: string;
  children: any;
  info?: string;
  url?: string;
  style?: object;
};

export const Panel: React.FC<PanelProps> = ({ title, children, info, url, style }) => (
  <Card style={{ marginBottom: 20, ...style }}>
    <Card.Title style={{ background: "#000091", color: "white", padding: 10 }}>
      {title}
      {info && (
        <span
          style={{ fontWeight: "normal", fontSize: "0.6em", marginLeft: 10 }}
        >
          {info}
        </span>
      )}
      {url && (
        <a
          style={{ color: "white", float: "right" }}
          href={url}
          target="_blank"
          rel="noreferrer noopener"
        >
          <ExternalLink />
        </a>
      )}
    </Card.Title>
    <Card.Body>{children}</Card.Body>
  </Card>
);
