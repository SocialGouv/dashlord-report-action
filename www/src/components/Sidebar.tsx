import * as React from "react";

import { Nav } from "react-bootstrap";
import { NavLink } from "react-router-dom";
import { Home, BarChart, Search } from "react-feather";
import uniq from "lodash.uniq"

import { smallUrl, sortByKey } from "../utils";

type SidebarProps = {
  report: any;
};

export const Sidebar: React.FC<SidebarProps> = ({ report }) => {
  const sortedReport = report.sort(sortByKey("url"));
  const categories = uniq(sortedReport.filter((u: any) => u.category).map((u: any) => u.category)).sort() as string[];
  return (
    <Nav className="col-md-3 col-lg-2 d-md-block bg-light sidebar collapse">
      <div className="sidebar-sticky pt-3">
        <ul className="nav flex-column">
          <li className="nav-item">
            <NavLink
              to="/"
              className="nav-link"
              activeClassName="active"
              exact={true}
            >
              <Home size={16} style={{ marginTop: -5, marginRight: 5 }} />
              Accueil
            </NavLink>
            <NavLink
              to="/dashboard"
              className="nav-link"
              activeClassName="active"
              exact={true}
            >
              <BarChart size={16} style={{ marginTop: -5, marginRight: 5 }} />
              Tableau de bord
            </NavLink>
          </li>
        </ul>

        {categories.length && <React.Fragment>
          <h6 className="sidebar-heading d-flex justify-content-between align-items-center px-3 mt-4 mb-1 text-muted">
            <span>Cat√©gories</span>
          </h6>
          <ul className="nav flex-column">
            {categories.map((category: string) => (
              <li
                className="nav-item"
                key={category}
                style={{
                  whiteSpace: "pre",
                }}
              >
                <NavLink
                  className="nav-link"
                  to={`/category/${category}`}
                  activeClassName="active"
                  exact={true}
                >
                  <BarChart size={16} style={{ marginTop: -5, marginRight: 5 }} />
                  {category}
                </NavLink>
              </li>
            ))}
          </ul>
        </React.Fragment> || null}


        <h6 className="sidebar-heading d-flex justify-content-between align-items-center px-3 mt-4 mb-1 text-muted">
          <span>Urls</span>
        </h6>

        <ul className="nav flex-column">
          {sortedReport.map((url: any) => (
            <li
              className="nav-item"
              key={url.url}
              style={{
                whiteSpace: "pre",
              }}
            >
              <NavLink
                className="nav-link"
                to={`/url/${url.url}`}
                activeClassName="active"
                exact={true}
              >
                <Search size={16} style={{ marginTop: -5, marginRight: 5 }} />
                {smallUrl(url.url)}
              </NavLink>
            </li>
          ))}
        </ul>
      </div>
    </Nav>
  );
};
