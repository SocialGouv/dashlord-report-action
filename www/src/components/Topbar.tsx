import * as React from "react";
import { Cloud } from "react-feather";
import { Navbar } from "react-bootstrap";

import { Link } from "react-router-dom";

export const Topbar = () => (
  <Navbar
    variant="dark"
    bg="dark"
    sticky="top"
    expand="lg"
    className="p-0 flex-md-nowrap shadow"
  >
    <Link to="/">
      <div
        className="px-3"
        style={{ height: 30, fontSize: "1.3rem", color: "var(--white)" }}
      >
        <Cloud size={16} style={{ marginTop: -5, marginRight: 10 }} />
        DashLord
      </div>
    </Link>
    {/*<FormControl
      type="text"
      placeholder="Search url"
      className=" form-control-dark w-100"
    />*/}
  </Navbar>
);
