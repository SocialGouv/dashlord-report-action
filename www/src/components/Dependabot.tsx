import * as React from "react";

import { Table, Badge } from "react-bootstrap";
import { getLastUrlSegment } from "../utils";

import { Panel } from "./Panel";

const orderBySeverity = (a: DependabotNode, b: DependabotNode) => {
  // high criticity first
  const severities = new Map();
  severities.set("CRITICAL", 3);
  severities.set("HIGH", 2);
  severities.set("MODERATE", 1);
  severities.set("LOW", 0);
  return severities.get(b.securityVulnerability.severity) - severities.get(a.securityVulnerability.severity);
};

const DependabotBadge = (node: DependabotNode) => {
  const severity = node.securityVulnerability.severity;
  const variant =
    severity === "LOW"
      ? "info"
      : severity === "MODERATE"
        ? "warning"
        : severity === "HIGH"
          ? "danger"
          : severity === "CRITICAL"
            ? "danger"
            : "info";
  return (
    <Badge className="w-100" variant={variant}>
      {severity}
    </Badge>
  );
};

type DependabotProps = { data: DependabotRepositoryVulnerabilities; url: string };

export const Dependabot: React.FC<DependabotProps> = ({ data, url }) => {
  console.log("data=" + JSON.stringify(data));
  const nodes =
    data && data.vulnerabilityAlerts.totalCount > 0 ? data.vulnerabilityAlerts.nodes : [];
  data.vulnerabilityAlerts.nodes.sort(orderBySeverity);
  return (
    (data.vulnerabilityAlerts.totalCount > 0 && (
      <Panel
        title="Dependabot"
        url={data.url}
        info={"Scan des vulnérabiliés du dépôt Github " + data.url}
      >
        <Table striped bordered hover>
          <thead>
            <tr>
              <th style={{ width: 100 }} className="text-center">
                Sévérité
              </th>
              <th>Dépendance vulnérable</th>
              <th>Vulnérabilités</th>
            </tr>
          </thead>
          <tbody>
            {nodes.map((node, i: number) => {
              return (
                <tr key={node.securityVulnerability.package.name + i}>
                  <td className="text-center">
                    <DependabotBadge {...node} />
                  </td>
                  <td>{node.securityVulnerability.package.name}</td>
                  <td>{node.securityVulnerability.advisory.references.map((reference, i: number) => {
                    return (<p key={getLastUrlSegment(reference.url) + i} ><a href={reference.url}>{getLastUrlSegment(reference.url)}</a><br /></p>);
                  })}</td>
                </tr>
              );
            })}
          </tbody>
        </Table>
      </Panel>
    )) ||
    null
  );
};