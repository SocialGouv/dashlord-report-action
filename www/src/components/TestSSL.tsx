import * as React from "react";

import { Row, Col, Alert, Table, Badge } from "react-bootstrap";
import { Info } from "react-feather";

import { Panel } from "./Panel";
import { Grade } from "./Grade";
import { cpuUsage } from "node:process";


const sortByKey = (key: string) => (a: any, b: any) => {
  if (a[key] > b[key]) {
    return 1;
  } else if (a[key] < b[key]) {
    return -1;
  }
  return 0;
};

//const grades = ["A+", "A", "B", "C", "D", "E", "F"]

const severities = ["INFO", "OK", "LOW", "MEDIUM", "HIGH", "CRITICAL"];
const getSeverityValue = (severity: string) => severities.indexOf(severity);

type SSLBadgeProps = { row: any; };

const SSLBadge: React.FC<SSLBadgeProps> = ({ row }) => {
  const severity = row.severity;
  console.log(row)
  const variant =
    (severity === "INFO" || severity === "OK" || severity === "LOW")
      ? "info"
      : severity === "MEDIUM"
        ? "warning"
        : "danger"
  return (
    <Badge className="w-100" variant={variant}>
      {row.severity}
    </Badge>
  );
};

type SSLProps = { data: any; url: string };

export const TestSSL: React.FC<SSLProps> = ({ data, url }) => {
  const grade =
    data && data.find((entry: any) => entry.id === "overall_grade").finding;
  const results = data.map((row: any) => ({
    ...row,
    severity_value: getSeverityValue(row.severity),
  }));
  results.sort(sortByKey("severity_value")).reverse();
  const capReasons = data.filter((entry: any) => entry.id.indexOf('grade_cap_reason_') === 0).reverse();
  return (
    data &&
    data.length && (
      <Panel title="SSL" info="Informations collectÃ©es via SSLlabs" url={url}>
        <Row>
          <Col>
            <h3>
              Scan Summary : <Grade small grade={grade} />
            </h3>
            <br />
            {capReasons.map((reason: any) => {
              return <Alert key={reason.id} variant="info"><Info style={{ marginRight: 5 }} />{reason.finding}</Alert>
            })}
          </Col>
        </Row>
      </Panel>
    )
  );
};
