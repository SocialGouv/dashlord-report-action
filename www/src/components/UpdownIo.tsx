import * as React from "react";
import { Row, Col, Alert, Card } from "react-bootstrap";
import { format } from "date-fns";
import frLocale from "date-fns/locale/fr";

import { Panel } from "./Panel";
import { Gauge } from "./Gauge";
import { smallUrl } from "../utils";

type UpDownIoProps = { data: UpDownReport; url: string };

export const UpdownIo: React.FC<UpDownIoProps> = ({ data, url }) => {
  const urlUpdownio = (data && `https://updown.io/${data.token}`) || null;

  return (
    (urlUpdownio && smallUrl(data.url) === smallUrl(url) && (
      <Panel
        title="Temps de réponse"
        info="Informations collectées par updown.io"
        url={urlUpdownio}
      >
        <Row>
          <Col xs={12} md={6} className="text-center mb-3">
            <Card>
              <Gauge
                width={100}
                height={20}
                value={data.uptime * 100}
                minValue={0}
                maxValue={100}
                animationSpeed={32}
              />
              <Card.Body>
                <Card.Title>
                  Taux de disponibilité sur un mois glissant
                </Card.Title>
                <Card.Title style={{ fontSize: "2rem", fontWeight: "bold" }}>
                  {data.uptime + "%"}
                </Card.Title>
              </Card.Body>
            </Card>
          </Col>
          <Col xs={12} md={6} className="text-center mb-3">
            {data.ssl && (
              <Card>
                <Card.Body>
                  <Card.Title>
                    Certificat TLS {data.ssl.valid ? "valide" : "expiré"}
                  </Card.Title>
                  <Card.Title style={{ fontSize: "2rem", fontWeight: "bold" }}>
                    expire le{" "}
                    {format(new Date(data.ssl.expires_at), "dd/MM/yyyy", {
                      locale: frLocale,
                    })}
                  </Card.Title>
                </Card.Body>
              </Card>
            )}
          </Col>
        </Row>
      </Panel>
    )) || (
      <Panel
        title="Temps de réponse"
        info="Informations collectées par updown.io"
      >
        <Alert variant="success">Aucune donnée updown.io associée</Alert>
      </Panel>
    )
  );
};
