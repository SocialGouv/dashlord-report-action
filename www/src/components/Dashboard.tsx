import * as React from "react";
import { useState, useMemo } from "react";
import { Slash, Info, Search, AlertTriangle } from "react-feather";
import { Link } from "react-router-dom";
import Tooltip from "rc-tooltip";
import orderBy from "lodash.orderby";

import BaseTable, { AutoResizer, Column, SortOrder } from "react-base-table";

import { Grade } from "./Grade";
import { smallUrl, isToolEnabled, letterGradeValue } from "../utils";
import { getPerformanceScore } from "../lib/lighthouse/getPerformanceScore";
import { AccessibilityWarnings } from "../lib/lighthouse/AccessibilityWarnings";
import { apdexToGrade } from "./UpdownIo";

import "react-base-table/styles.css";
import "rc-tooltip/assets/bootstrap.css";

type DashboardProps = { report: DashLordReport };

const remap = (value: number, x1: number, y1: number, x2: number, y2: number) =>
  ((value - x1) * (y2 - x2)) / (y1 - x1) + x2;

const scoreToGrade = (score: number) => {
  const grades = "A,B,C,D,E,F".split(",");

  const newGrade = Math.min(
    grades.length - 1,
    Math.floor(remap(score, 0, 1, 0, 6))
  );

  return grades[newGrade];
};

const IconUnknown = () => <Slash size={20} />;

const getGradeTrackers = (count: number) => {
  return count > 10 ? "F" : count > 2 ? "C" : count > 0 ? "B" : "A";
};

const getGradeCookies = (count: number) => {
  return count > 10
    ? "F"
    : count > 5
    ? "E"
    : count > 2
    ? "C"
    : count > 0
    ? "B"
    : "A";
};

const getGradeUpdownio = (uptime: number) => {
  return uptime > 0.99
    ? "A"
    : uptime > 0.98
    ? "B"
    : uptime > 0.97
    ? "C"
    : uptime > 0.96
    ? "D"
    : uptime > 0.95
    ? "E"
    : "F";
};

const getGradeOpenPorts = (count: number) => {
  return count > 2
    ? "F"
    : "A";
};

const getDependabotNodeGrade = (nodes: DependabotNode[]) => {
  return nodes.filter(
    (a) =>
      a.securityVulnerability.severity === "CRITICAL" ||
      a.securityVulnerability.severity === "HIGH"
  ).length
    ? "F"
    : nodes.length
    ? "B"
    : "A";
};

type ColumnHeaderProps = {
  title: string;
  info: string;
  warning?: React.ReactNode;
};

const ColumnHeader: React.FC<ColumnHeaderProps> = ({
  title,
  info,
  warning,
}) => (
  <div style={{ textAlign: "center" }}>
    <span style={{ fontSize: "0.9em" }}>
      {title}
      <br />
      <Tooltip
        placement="bottom"
        trigger={["hover"]}
        overlay={<div style={{ maxWidth: 300 }}>{info}</div>}
      >
        <Info size={16} style={{ cursor: "pointer" }} />
      </Tooltip>
    </span>

    {warning && (
      <Tooltip
        placement="bottom"
        trigger={["hover"]}
        overlay={<div style={{ maxWidth: 300 }}>{warning}</div>}
      >
        <AlertTriangle
          size={16}
          style={{
            stroke: "var(--danger)",
            marginLeft: 5,
            cursor: "pointer",
          }}
        />
      </Tooltip>
    )}
  </div>
);

type BadgeProps = { report: UrlReport };
type LightHouseBadgeProps = BadgeProps & {
  category: LighthouseReportCategoryKey;
};

const LightHouseBadge: React.FC<LightHouseBadgeProps> = ({
  report,
  category,
}) => {
  const lhrCategories = report.lhr && report.lhr.categories;
  if (!report.lhr || !lhrCategories) {
    return <IconUnknown />;
  }
  // use custom scoring
  lhrCategories["performance"].score = getPerformanceScore(report.lhr);

  const value =
    lhrCategories &&
    lhrCategories[category] &&
    (lhrCategories[category].score as number);
  return (
    <Grade
      small
      grade={scoreToGrade(1 - value)}
      label={(value * 100).toFixed() + " %"}
    />
  );
};

const SSLBadge: React.FC<BadgeProps> = ({ report }) => {
  const overallGrade =
    report.testssl &&
    report.testssl.find((entry) => entry.id === "overall_grade");
  const value = overallGrade && overallGrade.finding;
  if (!value) {
    return <IconUnknown />;
  }
  return <Grade small grade={value} />;
};

const HTTPBadge: React.FC<BadgeProps> = ({ report }) => {
  const value = report.http && report.http.grade;
  if (!value) {
    return <IconUnknown />;
  }
  return <Grade small grade={value} />;
};

const ThirdPartiesTrackersBadge: React.FC<BadgeProps> = ({ report }) => {
  if (!report.thirdparties) {
    return <IconUnknown />;
  }
  const trackersCount =
    (report.thirdparties &&
      report.thirdparties.trackers &&
      report.thirdparties.trackers.length) ||
    0;
  const trackersGrade = getGradeTrackers(trackersCount);
  return <Grade small grade={trackersGrade} label={trackersCount} />;
};

const ThirdPartiesCookiesBadge: React.FC<BadgeProps> = ({ report }) => {
  if (!report.thirdparties) {
    return <IconUnknown />;
  }
  const cookiesCount =
    (report.thirdparties &&
      report.thirdparties.cookies &&
      report.thirdparties.cookies.length) ||
    0;
  const cookiesGrade = getGradeCookies(cookiesCount);
  return <Grade small grade={cookiesGrade} label={cookiesCount} />;
};

const getNmapOpenPortGrade = (vulnerabilities: NmapVulnerability[]) => {
  return vulnerabilities.filter(
    (a) => a.is_exploit && Number.parseFloat(a.cvss) > 7
  ).length
    ? "F"
    : vulnerabilities.length
    ? "B"
    : "A";
};
const NmapBadge: React.FC<BadgeProps> = ({ report }) => {
  const value = report.nmap && report.nmap.grade;
  if (!value) {
    return <IconUnknown />;
  }
  return <Grade small grade={value} />;
};

const NmapOpenPortsBadge: React.FC<BadgeProps> = ({ report }) => {
  const openPortsCount =
    report.nmap && report.nmap.open_ports && report.nmap.open_ports.length;
  if (openPortsCount === undefined || openPortsCount === null) {
    return <IconUnknown />;
  }

  const openPortsGrade = getGradeOpenPorts(openPortsCount);
  return <Grade small grade={openPortsGrade} label={openPortsCount} />;
};

const DependabotBadge: React.FC<BadgeProps> = ({ report }) => {
  if (!report.dependabot) {
    return <IconUnknown />;
  }

  // dependabot
  const dependabotCount =
    report.dependabot &&
    report.dependabot
      .filter(Boolean)
      .map((repo) => repo.vulnerabilityAlerts.totalCount)
      .reduce((prev, curr) => prev + curr, 0);
  const maxGrade = (a: "F" | "B" | "A", b: "F" | "B" | "A") => {
    const grades = new Map();
    grades.set("F", 3);
    grades.set("B", 2);
    grades.set("A", 1);
    const orders = new Map();
    orders.set(3, "F");
    orders.set(2, "B");
    orders.set(1, "A");
    return orders.get(Math.max(grades.get(a), grades.get(b)));
  };
  const grades =
    report.dependabot &&
    report.dependabot
      .filter(Boolean)
      .map((repo) => getDependabotNodeGrade(repo.vulnerabilityAlerts.nodes));

  if (!grades.length) {
    return <IconUnknown />;
  }

  const dependabotGrade = grades.reduce(maxGrade);
  return <Grade small grade={dependabotGrade} label={dependabotCount} />;
};

const getCodescanAlertGrade = (alerts: CodescanAlert[]) => {
  return alerts.filter((a) => a.rule.severity === "error").length
    ? "F"
    : alerts.length
    ? "B"
    : "A";
};

const CodescanBadge: React.FC<BadgeProps> = ({ report }) => {
  if (!report.codescan) {
    return <IconUnknown />;
  }

  // codescan
  const codescanCount =
    report.codescan &&
    report.codescan
      .filter(Boolean)
      .map((repo) => (repo ? (repo.alerts ? repo.alerts.length : 0) : 0))
      .reduce((prev, curr) => prev + curr, 0);
  const maxGrade = (a: "F" | "B" | "A", b: "F" | "B" | "A") => {
    const grades = new Map();
    grades.set("F", 3);
    grades.set("B", 2);
    grades.set("A", 1);
    const orders = new Map();
    orders.set(3, "F");
    orders.set(2, "B");
    orders.set(1, "A");
    return orders.get(Math.max(grades.get(a), grades.get(b)));
  };
  const grades =
    report.codescan &&
    report.codescan
      .filter(Boolean)
      .map((repo) =>
        repo ? (repo.alerts ? getCodescanAlertGrade(repo.alerts) : "A") : "A"
      );

  if (!grades.length) {
    return <IconUnknown />;
  }

  const codescanGrade = grades.reduce(maxGrade);
  return <Grade small grade={codescanGrade} label={codescanCount} />;
};

const UpDownIoUptimeBadge: React.FC<BadgeProps> = ({ report }) => {
  if (!report.updownio) {
    return <IconUnknown />;
  }
  const updownio = report.updownio && report.updownio.uptime;
  const updownioGrade = getGradeUpdownio(updownio);
  return (
    <Grade small grade={updownioGrade} label={updownio.toFixed() + " %"} />
  );
};

const UpDownIoApDexBadge: React.FC<BadgeProps> = ({ report }) => {
  const apdex =
    report.updownio && report.updownio.metrics && report.updownio.metrics.apdex;
  if (apdex === undefined || apdex === null) {
    return <IconUnknown />;
  }

  const updownioGrade = apdexToGrade(apdex);
  return <Grade small grade={updownioGrade} label={apdex} />;
};

type SortState = {
  key: string;
  order: SortOrder;
  column: { [column: string]: string };
};

const defaultSort = {
  key: "url",
  order: "asc",
  column: { dataKey: "url" },
} as SortState;

export const Dashboard: React.FC<DashboardProps> = ({ report }) => {
  const [sortBy, setSortBy] = useState(defaultSort);

  const onColumnSort = (column: any) => {
    setSortBy(column);
  };

  const sortedReport = useMemo(() => {
    const getSortedRows = (rows: any) => {
      return orderBy(
        rows,
        (row) => {
          if (sortBy.column.dataGetter) {
            //@ts-expect-error
            return sortBy.column.dataGetter({ rowData: row });
          } else if (sortBy.column.key) {
            return row[sortBy.column.key];
          }
        },
        sortBy.order
      );
    };

    return getSortedRows(report);
  }, [sortBy, report]);

  const defaultColumnProps = {
    width: 120,
    sortable: true,
    align: "center",
  } as {
    width: number;
    sortable: boolean;
    align: "center" | "left" | "right";
  };

  return (
    <div style={{ width: "100%", height: "calc(100vh - 30px)" }}>
      <AutoResizer>
        {({ width, height }) => (
          <BaseTable
            data={sortedReport}
            width={width}
            height={height}
            sortBy={sortBy}
            onColumnSort={onColumnSort}
          >
            <Column
              key="url"
              title="url"
              sortable={true}
              width={300}
              flexGrow={1}
              cellRenderer={({ rowData }) => (
                <Link
                  to={`/url/${encodeURIComponent((rowData as UrlReport).url)}`}
                >
                  <Search size={16} />
                  &nbsp;{smallUrl((rowData as UrlReport).url)}
                </Link>
              )}
            />
            {isToolEnabled("lighthouse") && (
              <Column
                {...defaultColumnProps}
                key="accessibility"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return (
                    (report &&
                      report.lhr &&
                      report.lhr.categories.accessibility.score) ||
                    0
                  );
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="Accessibilité"
                    info="Bonnes pratiques en matière d'accessibilité web (LightHouse)"
                    warning={<AccessibilityWarnings />}
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <LightHouseBadge
                    report={rowData as UrlReport}
                    category="accessibility"
                  />
                )}
              />
            )}

            {isToolEnabled("lighthouse") && (
              <Column
                {...defaultColumnProps}
                key="performance"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return (report.lhr && getPerformanceScore(report.lhr)) || 0;
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="Performance"
                    info="Performances de chargement des pages web (LightHouse)"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <LightHouseBadge
                    report={rowData as UrlReport}
                    category="performance"
                  />
                )}
              />
            )}

            {isToolEnabled("lighthouse") && (
              <Column
                {...defaultColumnProps}
                key="seo"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return (
                    (report && report.lhr && report.lhr.categories.seo.score) ||
                    0
                  );
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="SEO"
                    info="Bonnes pratiques en matière de référencement naturel (LightHouse)"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <LightHouseBadge
                    report={rowData as UrlReport}
                    category="seo"
                  />
                )}
              />
            )}

            {isToolEnabled("testssl") && (
              <Column
                {...defaultColumnProps}
                key="ssl"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  const overallGrade =
                    report.testssl &&
                    report.testssl.find(
                      (entry) => entry.id === "overall_grade"
                    );
                  return overallGrade && letterGradeValue(overallGrade.finding);
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="SSL"
                    info="Niveau de confiance du certificat SSL (testssl.sh)"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <SSLBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("http") && (
              <Column
                {...defaultColumnProps}
                key="http"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return report.http && letterGradeValue(report.http.grade);
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="HTTP"
                    info="Bonnes pratiques de configuration HTTP (Mozilla observatory)"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <HTTPBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("updownio") && (
              <Column
                {...defaultColumnProps}
                key="updownio"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return report.updownio && report.updownio.uptime;
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="Disponibilité"
                    info="Disponibilité du service (updown.io)"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <UpDownIoUptimeBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("updownio") && (
              <Column
                {...defaultColumnProps}
                key="updownio2"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return (
                    report.updownio &&
                    report.updownio.metrics &&
                    report.updownio.metrics.apdex
                  );
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="apdex"
                    info="Apdex: Application Performance Index : indice de satisfaction des attentes de performance (updown.io)"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <UpDownIoApDexBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("dependabot") && (
              <Column
                {...defaultColumnProps}
                key="dependabot"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  const dependabotCount =
                    report.dependabot &&
                    report.dependabot
                      .filter(Boolean)
                      .map((repo) => repo.vulnerabilityAlerts.totalCount)
                      .reduce((prev, curr) => prev + curr, 0);
                  return dependabotCount;
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="Vulnérabilités"
                    info="Vulnérabilités applicatives detectées dans les dépendances du code (dependabot)"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <DependabotBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("nmap") && (
              <Column
                {...defaultColumnProps}
                key="nmap"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return report.nmap && letterGradeValue(report.nmap.grade);
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="Nmap"
                    info="Vulnérabilités réseau detectées par Nmap"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <NmapBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("nmap") && (
              <Column
                {...defaultColumnProps}
                key="nmap2"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return report.nmap && report.nmap.open_ports.length > 0;
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="Ports ouverts"
                    info="Ports TCP ouverts détectés par nmap"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <NmapOpenPortsBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("codescan") && (
              <Column
                {...defaultColumnProps}
                key="codescan"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  const codescanCount =
                    report.codescan &&
                    report.codescan
                      .filter(Boolean)
                      .map((repo) =>
                        repo ? (repo.alerts ? repo.alerts.length : 0) : 0
                      )
                      .reduce((prev, curr) => prev + curr, 0);
                  return codescanCount;
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="CodeQL"
                    info="Potentielles vulnérabilités ou erreurs detectées dans les codes sources (codescan)"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <CodescanBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("thirdparties") && (
              <Column
                {...defaultColumnProps}
                key="trackers"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return (
                    (report.thirdparties &&
                      report.thirdparties.trackers &&
                      report.thirdparties.trackers.length) ||
                    0
                  );
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="Trackers"
                    info="Nombre de scripts externes détectés"
                    warning={
                      <div>
                        Certains scripts externes légitimes peuvent être
                        considérés comme trackers.
                      </div>
                    }
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <ThirdPartiesTrackersBadge report={rowData as UrlReport} />
                )}
              />
            )}

            {isToolEnabled("thirdparties") && (
              <Column
                {...defaultColumnProps}
                key="cookies"
                dataGetter={({ rowData }) => {
                  const report = rowData as UrlReport;
                  return (
                    (report.thirdparties &&
                      report.thirdparties.cookies &&
                      report.thirdparties.cookies.length) ||
                    0
                  );
                }}
                headerRenderer={() => (
                  <ColumnHeader
                    title="Cookies"
                    info="Nombre de cookies présents"
                  />
                )}
                cellRenderer={({ rowData }) => (
                  <ThirdPartiesCookiesBadge report={rowData as UrlReport} />
                )}
              />
            )}
          </BaseTable>
        )}
      </AutoResizer>
    </div>
  );
};
