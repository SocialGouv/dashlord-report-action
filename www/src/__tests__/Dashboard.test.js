import * as React from "react";
import * as renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";

import { Dashboard } from "../components/Dashboard";

import report from "../report.json";

it("Should render empty Dashboard", () => {
  const props = {};
  const tree = renderer.create(<Dashboard {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Should render full Dashboard", () => {
  const props = { report };
  const tree = renderer
    .create(
      <MemoryRouter>
        <Dashboard {...props} />{" "}
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe("Tools config", () => {
  beforeEach(() => {
    jest.mock("../config.json", () => ({
      tools: ["http", "zap"],
    }));
  });

  it("Should render Dashboard with limited tools", () => {
    const props = { report };
    const tree = renderer
      .create(
        <MemoryRouter>
          <Dashboard {...props} />{" "}
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
