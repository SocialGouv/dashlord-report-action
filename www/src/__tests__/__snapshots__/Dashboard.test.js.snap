// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render empty Dashboard 1`] = `
<div
  style={
    Object {
      "height": "calc(100vh - 30px)",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "height": 0,
        "overflow": "visible",
        "width": 0,
      }
    }
  />
</div>
`;

exports[`Should render full Dashboard 1`] = `
Array [
  <div
    style={
      Object {
        "height": "calc(100vh - 30px)",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    />
  </div>,
  " ",
]
`;

exports[`Tools config Should render Dashboard with limited tools 1`] = `
Array [
  <div
    style={
      Object {
        "height": "calc(100vh - 30px)",
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    />
  </div>,
  " ",
]
`;
